opendc-experiments/opendc-m3sa/src/main/python/models/MultiModel.py

```diff
@@ -0,0 +1,501 @@
+
+        if self.unit_scaling == 1:
+            return given_metric
+
+        for i in range(len(scaling_factors)):
+            if self.unit_scaling == scaling_factors[i]:
+                self.unit_scaling = 10 ** self.unit_scaling
+                result = prefixes[i] + given_metric
+                return result
+
+    def set_paths(self):
+        """
+        Configures and initializes the directory paths for output and analysis based on the base directory provided.
+        This method sets paths for the raw output and detailed analysis results, ensuring directories are created if
+        they do not already exist, and prepares a base file for capturing analytical summaries.
+
+        :return: None
+        :side effect: Creates necessary directories and files for output and analysis.
+        """
+        self.output_folder_path = os.getcwd() + "/" + self.path
+        self.raw_output_path = os.getcwd() + "/" + self.path + "/raw-output"
+        self.analysis_file_path = os.getcwd() + "/" + self.path + "/simulation-analysis/"
+        os.makedirs(self.analysis_file_path, exist_ok=True)
+        self.analysis_file_path = os.path.join(self.analysis_file_path, "analysis.txt")
+        if not os.path.exists(self.analysis_file_path):
+            with open(self.analysis_file_path, "w") as f:
+                f.write("Analysis file created.\n")
+
+    def init_models(self):
+        """
+        Initializes models from the simulation output stored in Parquet files. This method reads each Parquet file,
+        processes the relevant data, and initializes Model instances which are stored in the model list.
+
+        :return: None
+        :raise ValueError: If the unit scaling has not been set prior to model initialization.
+        """
+        model_id = 0
+
+        for simulation_folder in os.listdir(self.raw_output_path):
+            if simulation_folder == "metamodel":
+                continue
+            path_of_parquet_file = f"{self.raw_output_path}/{simulation_folder}/seed={self.seed}/{SIMULATION_DATA_FILE}.parquet"
+            parquet_file = pq.read_table(path_of_parquet_file).to_pandas()
+            raw = parquet_file.select_dtypes(include=[np.number]).groupby("timestamp")
+            raw = raw[self.metric].sum().values
+
+            if self.unit_scaling is None:
+                raise ValueError("Unit scaling factor is not set. Please ensure it is set correctly.")
+
+            raw = np.divide(raw, self.unit_scaling)
+
+            if self.user_input["samples_per_minute"] > 0:
+                MINUTES_IN_DAY = 1440
+                self.workload_time = len(raw) * self.user_input["samples_per_minute"] / MINUTES_IN_DAY
+
+            model = Model(raw_sim_data=raw, id=model_id, path=self.output_folder_path)
+            self.models.append(model)
+            model_id += 1
+
+        self.max_model_len = min([len(model.raw_sim_data) for model in self.models])
+
+    def compute_windowed_aggregation(self):
```

Comment:
Pandas or NumPy should have functions that do this for you. 
There will always be more performant than what you can write in Python

Reply:
oh, I see. How to adapt this?